egui/src/widgets/plotting/items.rs

```diff
@@ -0,0 +1,214 @@
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------
+
+enum Generator {
+    /// Describes a function y = f(x) with an optional range for x and a number of points.
+    Explicit(Box<dyn Fn(f64) -> f64>, Option<RangeInclusive<f64>>, usize),
+    /// Describes a function (x,y) = f(t) with a range for t and a number of points.
+    Parametric(Box<dyn Fn(f64) -> (f64, f64)>, RangeInclusive<f64>, usize),
+}
+
+// ----------------------------------------------------------------------------
+
+/// A series of values forming a path.
+pub struct Curve {
+    pub(crate) values: Vec<Value>,
+    generator: Option<Generator>,
+    pub(crate) bounds: Bounds,
+    pub(crate) stroke: Stroke,
+    pub(crate) name: String,
+}
+
+impl Curve {
+    fn empty() -> Self {
+        Self {
+            values: Vec::new(),
+            generator: None,
+            bounds: Bounds::NOTHING,
+            stroke: Stroke::new(2.0, Color32::TRANSPARENT),
+            name: Default::default(),
+        }
+    }
+
+    pub fn from_values(values: Vec<Value>) -> Self {
+        let mut bounds = Bounds::NOTHING;
+        for value in &values {
+            bounds.extend_with(value);
+        }
+        Self {
+            values,
+            bounds,
+            ..Self::empty()
+        }
+    }
+
+    pub fn from_values_iter(iter: impl Iterator<Item = Value>) -> Self {
+        Self::from_values(iter.collect())
+    }
+
+    /// Draw a curve based on a function `y=f(x)`, an optional range for x and the number of points.
+    pub fn from_explicit_callback(
+        function: impl Fn(f64) -> f64 + 'static,
+        range: Option<RangeInclusive<f64>>,
+        points: usize,
+    ) -> Self {
+        let generator = Generator::Explicit(Box::new(function), range, points);
+        Self {
+            generator: Some(generator),
+            ..Self::empty()
+        }
+    }
+
+    /// Draw a curve based on a function `(x,y)=f(t)`, a range for t and the number of points.
+    pub fn from_parametric_callback(
+        function: impl Fn(f64) -> (f64, f64) + 'static,
+        t_range: RangeInclusive<f64>,
+        points: usize,
+    ) -> Self {
+        let generator = Generator::Parametric(Box::new(function), t_range, points);
+        Self {
+            generator: Some(generator),
+            ..Self::empty()
+        }
+    }
+
+    /// Returns true if there are no data points available and there is no function to generate any.
+    pub(crate) fn no_data(&self) -> bool {
+        self.generator.is_none() && self.values.is_empty()
+    }
+
+    /// If initialized with a generator function, this will generate `n` evenly spaced points in the
+    /// given range.
+    pub(crate) fn generate_points(&mut self, mut x_range: RangeInclusive<f64>) {
+        let range_union = |range1: &RangeInclusive<f64>, range2: &RangeInclusive<f64>| {
+            range1.start().max(*range2.start())..=range1.end().min(*range2.end())
```

Comment:
This is the intersection; not the union

Reply:
This would also be better as `fn` instead of a lambda, since it doesn't capture anything.